### PR TITLE
Check program state before permitting linking

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Resolves: #2248 

## Description

Program state should be protected by the lock.
Linking should be blocked if code isn't loaded yet.

## Testing

CI/CD

## Documentation

No.

### CODE REVIEW

File: libs/execution_context/ebpf_program.c [hunk 1]
```diff
@@ -599,35 +608,44 @@ ebpf_program_associate_additional_map(ebpf_program_t* program, ebpf_map_t* map)
...
     program->count_of_maps = maps_count;
     for (index = 0; index < maps_count; index++) {
-        ebpf_object_acquire_reference((ebpf_core_object_t*)program_maps[index]);
+        ebpf_object_acquire_reference((ebpf_core_object_t*)program->maps[index]);
     }
+    ebpf_lock_unlock(&program->lock, state);
 
-    EBPF_RETURN_RESULT(EBPF_SUCCESS);
+Done:
+    if (program_maps) {
```
Comment (dthaler) [inline libs/execution_context/ebpf_program.c:644]: Remove `if` test to reduce number of code paths to test.  `ebpf_free` is safe to pass nullptr to.

File: libs/execution_context/ebpf_program.h [hunk 2]
```diff
@@ -81,18 +81,40 @@ extern "C"
...
+     * @retval EBPF_NO_MEMORY Unable to allocate resources.
+     */
+    _Must_inspect_result_ ebpf_result_t
+    ebpf_program_get_program_file_name(_In_ const ebpf_program_t* program, _Out_ ebpf_utf8_string_t* file_name);
+
+    /**
+     * @brief Get the original section name of the program.
+     *
+     * @param[in] program The program instance.
+     * @param section_name The section name of the program. Caller must free this.
```
Comment (dthaler) [inline libs/execution_context/ebpf_program.h:98]:
```suggestion
     * @param[out] section_name The section name of the program. Caller must free this.
```

File: libs/platform/ebpf_platform.h [hunk 3]
```diff
@@ -379,15 +387,15 @@ extern "C"
...
+        _IRQL_raises_(DISPATCH_LEVEL) ebpf_lock_state_t ebpf_lock_lock(_Inout_ const ebpf_lock_t* lock);
 
     /**
      * @brief Release exclusive access to the lock.
      * @param[in, out] lock Pointer to memory location that contains the lock.
      * @param[in] state The state returned from ebpf_lock_lock.
      */
     _Requires_lock_held_(*lock) _Releases_lock_(*lock) _IRQL_requires_(DISPATCH_LEVEL) void ebpf_lock_unlock(
-        _Inout_ ebpf_lock_t* lock, _IRQL_restores_ ebpf_lock_state_t state);
+        _Inout_ const ebpf_lock_t* lock, _IRQL_restores_ ebpf_lock_state_t state);
```
Comment (dthaler) [inline libs/platform/ebpf_platform.h:398]: How can it be Inout if it's const?

File: libs/platform/kernel/ebpf_platform_kernel.c [hunk 4]
```diff
@@ -382,15 +382,15 @@ ebpf_lock_destroy(_In_ _Post_invalid_ ebpf_lock_t* lock)
 }
 
 _Requires_lock_not_held_(*lock) _Acquires_lock_(*lock) _IRQL_requires_max_(DISPATCH_LEVEL) _IRQL_saves_
-    _IRQL_raises_(DISPATCH_LEVEL) ebpf_lock_state_t ebpf_lock_lock(_Inout_ ebpf_lock_t* lock)
+    _IRQL_raises_(DISPATCH_LEVEL) ebpf_lock_state_t ebpf_lock_lock(_Inout_ const ebpf_lock_t* lock)
 {
-    return KeAcquireSpinLockRaiseToDpc(lock);
+    return KeAcquireSpinLockRaiseToDpc((KSPIN_LOCK*)lock);
```
Comment (dthaler) [inline libs/platform/kernel/ebpf_platform_kernel.c:387]: This loses the const, so I don't think 385 should have "const"

File: libs/platform/user/ebpf_platform_user.cpp [hunk 5]
```diff
@@ -758,17 +758,17 @@ ebpf_lock_destroy(_In_ _Post_invalid_ ebpf_lock_t* lock)
 }
 
 _Requires_lock_not_held_(*lock) _Acquires_lock_(*lock) _IRQL_requires_max_(DISPATCH_LEVEL) _IRQL_saves_
-    _IRQL_raises_(DISPATCH_LEVEL) ebpf_lock_state_t ebpf_lock_lock(_Inout_ ebpf_lock_t* lock)
+    _IRQL_raises_(DISPATCH_LEVEL) ebpf_lock_state_t ebpf_lock_lock(_Inout_ const ebpf_lock_t* lock)
 {
-    AcquireSRWLockExclusive(reinterpret_cast<PSRWLOCK>(lock));
+    AcquireSRWLockExclusive(reinterpret_cast<PSRWLOCK>(const_cast<ebpf_lock_t*>(lock)));
```
Comment (dthaler) [inline libs/platform/user/ebpf_platform_user.cpp:763]: This also loses the const so I think line 761 should not have the const.

File: libs/platform/ebpf_platform.h (hunk 3 above)
Comment (Alan-Jowett) [reply libs/platform/ebpf_platform.h:398]: Not quite sure how to handle this case. There are paths where we need to read state from an object protected by a lock., but acquiring the lock is essentially a mutable operation. Unfortunately, C doesn't appear to have a way to mark the lock field as mutable, so either we cast at the ebpf_lock_lock call-site or we cast it internally.

**Review by saxena-anurag — DISMISSED** [review]

**Review by dthaler — APPROVED** [review]

**Review by dv-msft — APPROVED** [review]

**Pull request merged**